### ISSUE TITLE
[Console] Option -e prod is interpreted as a command when no command is specified

### ISSUE THREAD

**Issue by symcaster** [symfony/symfony#23343]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.3.2

I usually use the short command `bin/console` to list all available options. But on a production server, I have to set the env option to avoid default dev environment and missing debug vendors.

But, if I type only `bin/console -e prod` the command report the following error:

```
  [Symfony\Component\Console\Exception\CommandNotFoundException]  
  Command "prod" is not defined.                                  
  Did you mean this?                                              
      doctrine:ensure-production-settings
```

As soon as I set a command, there is no problem (for example `bin/console list -e prod`. The word `prod` is interpreted as the command instead of the value of the `-e` option.

**Comment by romain-pierre** [comment]
Hi, yes absolutely. In my use case it's more safe to rely on the command parameter. Obviously this is not a major bug. I will try to investigate in the next days.

**Comment by magnetik** [comment]
It's because of the `Application::getCommandName()` that returns the first argument as command name: https://github.com/symfony/symfony/blame/master/src/Symfony/Component/Console/Application.php#L832

**Comment by romain-pierre** [comment]
Suggest (also reminder for me): If an option have the mode VALUE_REQUIRED, the argument just after should be attached to its value, if not already an option.

**Comment by yceruto** [comment]
To avoid this issue you could try this syntax `bin/console -e=prod`?

**Comment by romain-pierre** [comment]
Yes, it's true. This is not a big issue, but the console does not behave as it should. In other cases, putting the value after a space is not a problem, so it should be also not here :)

**Comment by javiereguiluz** [comment]
Even if there are alternative solutions (`SYMFONY_ENV` var or `-e=prod`) I think @romain-pierre is right and we should consider this a bug.

**Comment by chalasr** [comment]
Indeed, I'll give it a try this week.

**Comment by Simperfit** [comment]
@chalasr did you already implemented it, or could I look into it ?

**Comment by HeahDude** [comment]
Isn't it somehow expected? I guess in such case `bin/console -- -e prod` should be used right?

**Comment by Simperfit** [comment]
@HeahDude IMO it should be working and return the available command in prod mod, like `kubectl` for exemple with the `-n (for --namespace)`

**Comment by chalasr** [comment]
@Simperfit I started to look into, the issue starts from https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/Input/ArgvInput.php#L259. It should consider a space as separator for option values, but it doesn't (considers `=` only). The right fix would be to check that the previous token starts with `-`, and if it does, check that the option is part of the definition and accepts a value. Problem is that at this point (of the application run process) the `--env` option is not yet part of the input definition, it is only after that `mergeApplicationDefinition` has been called, which happens later. 
I have to look deeper into, but if you want to give it a try feel free to do so! :)
Thanks for the ping.

**Comment by romain-pierre** [comment]
This is the only way I've found at this time to resolve this issue: https://github.com/romain-pierre/symfony/commit/e76a597fb8ef898bc27a62a146cd04f4a2d0b7bc

**Comment by Simperfit** [comment]
@romain-pierre Do you want to create the pull request ? You already did the feature, you may have to add test ;).

**Comment by romain-pierre** [comment]
I would like to have an advise from @chalasr before, because I don't know the impact of calling `bind()` twice.

**Comment by chalasr** [comment]
👍  for fixing `getFirstArgument()` and binding the input before checking for global options. 
We need to wrap the call it in a try-catch though, `bind()` would throw for any input argument that is not part of the application definition (i.e. the command args). 
PR with test case welcomed, should target 2.7

**Comment by issei-m** [comment]
Any news?

**Comment by fabpot** [comment]
For Symfony 4.1, I would recommend to deprecate both `-e` and `--no-debug` as they are redundant with the new env vars (`APP_ENV` and `APP_DEBUG`).

**Comment by Simperfit** [comment]
@romain-pierre Do you want to fix the behaviour for 2.7 to 4.0 ?
And for 4.1 we should deprecate it.

**Comment by chalasr** [comment]
@Simperfit Userland applications can have similar options, the behavior should not be deprecated (only some core global options should be).

**Comment by steve-todorov** [comment]
How exactly should I start the `server:start` command if I would like to test the `prod` environment? 

```
# SYMFONY_ENV=prod bin/console server:start 0.0.0.0:8000
 [ERROR] The document root directory must be either passed as first argument of the constructor or through the "docroot"
         input option.                                                                                                  

# SYMFONY_ENV=prod bin/console server:start 0.0.0.0:8000
 [ERROR] The document root directory must be either passed as first argument of the constructor or through the "docroot"
         input option.                                                                                                  
                                                                                                                        
# export SYMFONY_ENV=prod ; bin/console server:start 0.0.0.0:8000
 [ERROR] The document root directory must be either passed as first argument of the constructor or through the "docroot"
         input option.                                                                                                  
                                                                                                                        
# export SYMFONY_ENV=prod ; bin/console server:start 0.0.0.0:8000 --docroot ./web/
 [ERROR] The environment must be passed as second argument of the constructor.                                          

# SYMFONY_ENV=prod bin/console server:start 0.0.0.0:8000 --docroot ./web/
 [ERROR] The environment must be passed as second argument of the constructor.                                          

# SYMFONY_ENV=prod bin/console -e=prod server:start 0.0.0.0:8000 --docroot ./web/
 [ERROR] The environment must be passed as second argument of the constructor.  

# bin/console server:start 0.0.0.0:8000 --docroot ./web/ --env=prod                                                                                                                        
 [ERROR] The environment must be passed as second argument of the constructor.   
```

It doesn't seem like you could export the SYMFONY_ENV and then execute the symfony command at all?

**Comment by MrNicodemuz** [comment]
I'm also facing the same problem as @steve-todorov .. can't seem to use the web server bundle ..

**Comment by MrNicodemuz** [comment]
Managed to solve the problem. The problem for me was that I was mixing Symfony 3.4 and 4.0 bundles together. Surprisingly this was not caught by the package manager (composer) as a conflict.

After removing the distribution-bundle, all other bundles got correctly upgraded to correct versions.

**Comment by LinkingYou** [comment]
I have the same issue, that i can't set the environment through the command line option.

Works fine:
`php bin/console app:doit`

Works fine too:
`php bin/console app:doit --env=dev`

Fails with "There are no commands defined in the "app" namespace.":
`php bin/console app:doit --env=prod`

To set this by an environment variable is no option for me, because my command-line-tool should run twice on the same machine. One version in **dev** and the other in **prod**.

I am using symfony 4.1.3.

**Comment by stof** [comment]
@LinkingYou you can set env variables for each call too

**Comment by symcaster** [comment]
I think this issue can be closed since #28653

**Comment by nicolas-grekas** [comment]
> Options defined at the application level are meant to be available to all registered commands

Unless one opts in to a different meaning by using the proposed extra argument. I'd really prefer no hard coded special edge case btw.
